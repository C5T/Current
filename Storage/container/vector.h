/*******************************************************************************
The MIT License (MIT)

Copyright (c) 2015 Maxim Zhurovich <zhurovich@gmail.com>
          (c) 2015 Dmitry "Dima" Korolev <dmitry.korolev@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*******************************************************************************/

#ifndef CURRENT_STORAGE_CONTAINER_VECTOR_H
#define CURRENT_STORAGE_CONTAINER_VECTOR_H

#include "../base.h"

#include "../../TypeSystem/optional.h"

namespace current {
namespace storage {
namespace container {

struct CannotPopBackFromEmptyVectorException : Exception {};
typedef const CannotPopBackFromEmptyVectorException& CannotPopBackFromEmptyVector;

template <typename T, typename ADDER, typename DELETER>
class Vector {
 public:
  explicit Vector(MutationJournal& journal) : journal_(journal) {}

  bool Empty() const { return vector_.empty(); }
  size_t Size() const { return vector_.size(); }

  ImmutableOptional<T> operator[](size_t index) const {
    if (index < vector_.size()) {
      return ImmutableOptional<T>(FromBarePointer(), &vector_[index]);
    } else {
      return ImmutableOptional<T>(nullptr);
    }
  }

  void PushBack(const T& object) {
    journal_.LogMutation(ADDER(object), [this]() { vector_.pop_back(); });
    vector_.push_back(object);
  }

  void PopBack() {
    if (!vector_.empty()) {
      auto back_object = vector_.back();
      journal_.LogMutation(DELETER(back_object), [this, back_object]() { vector_.push_back(back_object); });
      vector_.pop_back();
    } else {
      CURRENT_THROW(CannotPopBackFromEmptyVectorException());
    }
  }

  void operator()(const ADDER& object) { vector_.push_back(object); }

  void operator()(const DELETER&) { vector_.pop_back(); }

 private:
  std::vector<T> vector_;
  MutationJournal& journal_;
};

}  // namespace container
}  // namespace storage
}  // namespace current

using current::storage::container::Vector;

#endif  // CURRENT_STORAGE_CONTAINER_VECTOR_H
