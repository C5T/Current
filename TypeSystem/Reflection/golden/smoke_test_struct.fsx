// fsharpi -r Newtonsoft.Json.dll current.fsx

open Newtonsoft.Json
let inline JSON o = JsonConvert.SerializeObject(o)
let inline ParseJSON (s : string) : 'T = JsonConvert.DeserializeObject<'T>(s)

type Primitives = {
  a : byte
  b : uint16
  c : uint32
  d : uint64
  e : sbyte
  f : int16
  g : int32
  h : int64
  i : char
  j : string
  k : float
  l : double
  m : bool
  n : int64  // microseconds.
  o : int64  // milliseconds.
}
type FullTest = {
  primitives : Primitives
  v1 : string array
  v2 : Primitives array
  p : string * Primitives
  o : Primitives option
}
