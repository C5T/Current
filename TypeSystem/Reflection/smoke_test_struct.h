#ifdef SMOKE_TEST_STRUCT_NAMESPACE

namespace SMOKE_TEST_STRUCT_NAMESPACE {

CURRENT_STRUCT(Primitives) {
  CURRENT_FIELD(a, uint8_t);
  CURRENT_FIELD(b, uint16_t);
  CURRENT_FIELD(c, uint32_t);
  CURRENT_FIELD(d, uint64_t);
  CURRENT_FIELD(e, int8_t);
  CURRENT_FIELD(f, int16_t);
  CURRENT_FIELD(g, int32_t);
  CURRENT_FIELD(h, int64_t);
  CURRENT_FIELD(i, char);
  CURRENT_FIELD(j, std::string);
  CURRENT_FIELD(k, float);
  CURRENT_FIELD(l, double);
  CURRENT_FIELD(m, bool);
  CURRENT_FIELD(n, std::chrono::microseconds);
  CURRENT_FIELD(o, std::chrono::milliseconds);
};

CURRENT_STRUCT(FullTest) {
  CURRENT_FIELD(primitives, Primitives);
  CURRENT_FIELD(v1, std::vector<std::string>);
  CURRENT_FIELD(v2, std::vector<Primitives>);
  CURRENT_FIELD(p, (std::pair<std::string, Primitives>));
  CURRENT_FIELD(o, Optional<Primitives>);
};

}  // namespace

#endif
