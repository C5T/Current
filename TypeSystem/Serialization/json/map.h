/*******************************************************************************
The MIT License (MIT)

Copyright (c) 2015 Dmitry "Dima" Korolev <dmitry.korolev@gmail.com>
          (c) 2016 Maxim Zhurovich <zhurovich@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*******************************************************************************/

#ifndef CURRENT_TYPE_SYSTEM_SERIALIZATION_JSON_MAP_H
#define CURRENT_TYPE_SYSTEM_SERIALIZATION_JSON_MAP_H

#include <type_traits>
#include <map>

#include "json.h"

#include "../../../bricks/template/enable_if.h"

namespace current {
namespace serialization {

template <class JSON_FORMAT, typename TK, typename TV, typename TC, typename TA>
struct SerializeImpl<json::JSONStringifier<JSON_FORMAT>, std::map<TK, TV, TC, TA>> {
  static void DoSerialize(json::JSONStringifier<JSON_FORMAT>& json_stringifier, const std::map<TK, TV, TC, TA>& value) {
    json_stringifier.Current().SetArray();
    for (const auto& element : value) {
      rapidjson::Value key_value_as_array;
      key_value_as_array.SetArray();
      rapidjson::Value populated_key;
      rapidjson::Value populated_value;
      json_stringifier.Inner(&populated_key, element.first);
      json_stringifier.Inner(&populated_value, element.second);
      key_value_as_array.PushBack(std::move(populated_key.Move()), json_stringifier.Allocator());
      key_value_as_array.PushBack(std::move(populated_value.Move()), json_stringifier.Allocator());
      json_stringifier.Current().PushBack(std::move(key_value_as_array.Move()), json_stringifier.Allocator());
    }
  }
};

template <class JSON_FORMAT, typename TV, typename TC, typename TA>
struct SerializeImpl<json::JSONStringifier<JSON_FORMAT>, std::map<std::string, TV, TC, TA>> {
  static void DoSerialize(json::JSONStringifier<JSON_FORMAT>& json_stringifier,
                          const std::map<std::string, TV, TC, TA>& value) {
    json_stringifier.Current().SetObject();
    for (const auto& element : value) {
      rapidjson::Value populated_value;
      json_stringifier.Inner(&populated_value, element.second);
      json_stringifier.Current().AddMember(
          rapidjson::StringRef(element.first), std::move(populated_value.Move()), json_stringifier.Allocator());
    }
  }
};

template <class JSON_FORMAT, typename TK, typename TV, typename TC, typename TA, class J>
struct DeserializeImpl<json::JSONParser<JSON_FORMAT>, std::map<TK, TV, TC, TA>, J> {
  template <typename K = TK>
  static std::enable_if_t<std::is_same<std::string, K>::value> DoDeserialize(json::JSONParser<JSON_FORMAT>& json_parser,
                                                                             std::map<TK, TV, TC, TA>& destination) {
    if (json_parser && json_parser.Current().IsObject()) {
      destination.clear();
      TK k;
      TV v;
      for (rapidjson::Value::MemberIterator cit = json_parser.Current().MemberBegin();
           cit != json_parser.Current().MemberEnd();
           ++cit) {
        json_parser.Inner(&cit->name, k);
        json_parser.Inner(&cit->value, v);
        destination.emplace(std::move(k), std::move(v));
      }
    } else if (!json::JSONPatchMode<J>::value || (json_parser && !json_parser.Current().IsObject())) {
      throw JSONSchemaException("map as object", json_parser);  // LCOV_EXCL_LINE
    }
  }

  template <typename K = TK>
  static std::enable_if_t<!std::is_same<std::string, K>::value> DoDeserialize(
      json::JSONParser<JSON_FORMAT>& json_parser, std::map<TK, TV, TC, TA>& destination) {
    if (json_parser && json_parser.Current().IsArray()) {
      destination.clear();
      for (rapidjson::Value::ValueIterator cit = json_parser.Current().Begin(); cit != json_parser.Current().End();
           ++cit) {
        if (!cit->IsArray()) {
          throw JSONSchemaException("map entry as array", json_parser);  // LCOV_EXCL_LINE
        }
        if (cit->Size() != 2u) {
          throw JSONSchemaException("map entry as array of two elements", json_parser);  // LCOV_EXCL_LINE
        }
        TK k;
        TV v;
        json_parser.Inner(&(*cit)[static_cast<rapidjson::SizeType>(0)], k);
        json_parser.Inner(&(*cit)[static_cast<rapidjson::SizeType>(1)], v);
        destination.emplace(std::move(k), std::move(v));
      }
    } else if (!json::JSONPatchMode<J>::value || (json_parser && !json_parser.Current().IsArray())) {
      throw JSONSchemaException("map as array", json_parser);  // LCOV_EXCL_LINE
    }
  }
};

namespace json {
template <typename K, typename V, typename CMP, typename ALLOC>
struct IsJSONSerializable<std::map<K, V, CMP, ALLOC>> {
  constexpr static bool value = IsJSONSerializable<K>::value && IsJSONSerializable<V>::value;
};
}  // namespace json

}  // namespace current::serialization
}  // namespace current

#endif  // CURRENT_TYPE_SYSTEM_SERIALIZATION_JSON_MAP_H
