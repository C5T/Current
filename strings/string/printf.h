#ifndef BRICKS_STRING_PRINTF_H
#define BRICKS_STRING_PRINTF_H

#include <string>
#include <cstdarg>

namespace bricks {
namespace string {

std::string Printf(const char *fmt, ...) {
  const int max_formatted_output_length = 1024 * 1024;
  static char buf[max_formatted_output_length + 1];
  va_list ap;
  va_start(ap, fmt);
  vsnprintf(buf, max_formatted_output_length, fmt, ap);
  va_end(ap);
  return buf;
}

}  // namespace string
}  // namespace bricks

#endif  // BRICKS_STRING_PRINTF_H
