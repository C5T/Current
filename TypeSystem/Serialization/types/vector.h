/*******************************************************************************
The MIT License (MIT)

Copyright (c) 2015 Dmitry "Dima" Korolev <dmitry.korolev@gmail.com>
          (c) 2016 Maxim Zhurovich <zhurovich@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*******************************************************************************/

#ifndef CURRENT_TYPE_SYSTEM_SERIALIZATION_TYPES_VECTOR_H
#define CURRENT_TYPE_SYSTEM_SERIALIZATION_TYPES_VECTOR_H

#include <vector>

#include "../base.h"

namespace current {
namespace serialization {
namespace json {

namespace save {

template <typename TT, typename TA, class J>
struct SaveIntoJSONImpl<std::vector<TT, TA>, J> {
  static bool Save(rapidjson::Value& destination,
                   rapidjson::Document::AllocatorType& allocator,
                   const std::vector<TT, TA>& value) {
    destination.SetArray();
    rapidjson::Value element_to_push;
    for (const auto& element : value) {
      SaveIntoJSONImpl<TT, J>::Save(element_to_push, allocator, element);
      destination.PushBack(element_to_push, allocator);
    }
    return true;
  }
};

}  // namespace save

namespace load {

template <typename TT, typename TA, class J>
struct LoadFromJSONImpl<std::vector<TT, TA>, J> {
  static void Load(rapidjson::Value* source, std::vector<TT, TA>& destination, const std::string& path) {
    if (source && source->IsArray()) {
      const size_t size = source->Size();
      destination.resize(size);
      for (rapidjson::SizeType i = 0; i < static_cast<rapidjson::SizeType>(size); ++i) {
        LoadFromJSONImpl<TT, J>::Load(&((*source)[i]), destination[i], path + '[' + std::to_string(i) + ']');
      }
    } else {
      throw JSONSchemaException("array", source, path);  // LCOV_EXCL_LINE
    }
  }
};

}  // namespace load
}  // namespace json

namespace binary {
namespace save {

template <typename TT, typename TA>
struct SaveIntoBinaryImpl<std::vector<TT, TA>> {
  static void Save(std::ostream& ostream, const std::vector<TT, TA>& value) {
    SaveSizeIntoBinary(ostream, value.size());
    for (const auto& element : value) {
      SaveIntoBinaryImpl<TT>::Save(ostream, element);
    }
  }
};

}  // namespace save

namespace load {

template <typename TT, typename TA>
struct LoadFromBinaryImpl<std::vector<TT, TA>> {
  static void Load(std::istream& istream, std::vector<TT, TA>& destination) {
    BINARY_FORMAT_SIZE_TYPE size = LoadSizeFromBinary(istream);
    destination.resize(static_cast<size_t>(size));
    for (size_t i = 0; i < static_cast<size_t>(size); ++i) {
      LoadFromBinaryImpl<TT>::Load(istream, destination[i]);
    }
  }
};

}  // namespace load
}  // namespace binary
}  // namespace serialization
}  // namespace current

#endif  // CURRENT_TYPE_SYSTEM_SERIALIZATION_TYPES_VECTOR_H
