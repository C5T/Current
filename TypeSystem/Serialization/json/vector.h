/*******************************************************************************
The MIT License (MIT)

Copyright (c) 2015 Dmitry "Dima" Korolev <dmitry.korolev@gmail.com>
          (c) 2016 Maxim Zhurovich <zhurovich@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*******************************************************************************/

#ifndef CURRENT_TYPE_SYSTEM_SERIALIZATION_JSON_VECTOR_H
#define CURRENT_TYPE_SYSTEM_SERIALIZATION_JSON_VECTOR_H

#include <vector>

#include "json.h"

namespace current {
namespace serialization {

template <class JSON_FORMAT, typename T>
struct SerializeImpl<json::JSONStringifier<JSON_FORMAT>, std::vector<T>> {
  static void DoSerialize(json::JSONStringifier<JSON_FORMAT>& json_stringifier, const std::vector<T>& value) {
    json_stringifier.Current().SetArray();
    for (const auto& element : value) {
      rapidjson::Value element_to_push;
      json_stringifier.Inner(&element_to_push, element);
      json_stringifier.Current().PushBack(std::move(element_to_push.Move()), json_stringifier.Allocator());
    }
  }
};

namespace json {
namespace load {

template <typename TT, typename TA, class J>
struct LoadFromJSONImpl<std::vector<TT, TA>, J> {
  static void Load(rapidjson::Value* source, std::vector<TT, TA>& destination, const std::string& path) {
    if (source && source->IsArray()) {
      const size_t size = source->Size();
      destination.resize(size);
      for (rapidjson::SizeType i = 0; i < static_cast<rapidjson::SizeType>(size); ++i) {
        LoadFromJSONImpl<TT, J>::Load(&((*source)[i]), destination[i], path + '[' + std::to_string(i) + ']');
      }
    } else if (!JSONPatchMode<J>::value || (source && !source->IsArray())) {
      throw JSONSchemaException("array", source, path);  // LCOV_EXCL_LINE
    }
  }
};

}  // namespace current::serialization::json::load
}  // namespace current::serialization::json
}  // namespace current::serialization
}  // namespace current

#endif  // CURRENT_TYPE_SYSTEM_SERIALIZATION_JSON_VECTOR_H
