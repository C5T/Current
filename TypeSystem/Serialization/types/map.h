/*******************************************************************************
The MIT License (MIT)

Copyright (c) 2015 Dmitry "Dima" Korolev <dmitry.korolev@gmail.com>
          (c) 2016 Maxim Zhurovich <zhurovich@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*******************************************************************************/

#ifndef CURRENT_TYPE_SYSTEM_SERIALIZATION_TYPES_MAP_H
#define CURRENT_TYPE_SYSTEM_SERIALIZATION_TYPES_MAP_H

#include <type_traits>
#include <map>

#include "../base.h"

#include "../../../Bricks/template/enable_if.h"

namespace current {
namespace serialization {
namespace json {
namespace save {

template <typename TK, typename TV, typename TC, typename TA, class J>
struct SaveIntoJSONImpl<std::map<TK, TV, TC, TA>, J> {
  static bool Save(rapidjson::Value& destination,
                   rapidjson::Document::AllocatorType& allocator,
                   const std::map<TK, TV, TC, TA>& value) {
    destination.SetArray();
    for (const auto& element : value) {
      rapidjson::Value key_value_as_array;
      key_value_as_array.SetArray();
      rapidjson::Value populated_key;
      rapidjson::Value populated_value;
      SaveIntoJSONImpl<TK, J>::Save(populated_key, allocator, element.first);
      SaveIntoJSONImpl<TV, J>::Save(populated_value, allocator, element.second);
      key_value_as_array.PushBack(populated_key, allocator);
      key_value_as_array.PushBack(populated_value, allocator);
      destination.PushBack(key_value_as_array, allocator);
    }
    return true;
  }
};

template <typename TV, typename TC, typename TA, class J>
struct SaveIntoJSONImpl<std::map<std::string, TV, TC, TA>, J> {
  static bool Save(rapidjson::Value& destination,
                   rapidjson::Document::AllocatorType& allocator,
                   const std::map<std::string, TV, TC, TA>& value) {
    destination.SetObject();
    for (const auto& element : value) {
      rapidjson::Value populated_value;
      SaveIntoJSONImpl<TV, J>::Save(populated_value, allocator, element.second);
      destination.AddMember(rapidjson::StringRef(element.first), populated_value, allocator);
    }
    return true;
  }
};

}  // namespace save

namespace load {

template <typename TK, typename TV, typename TC, typename TA, class J>
struct LoadFromJSONImpl<std::map<TK, TV, TC, TA>, J> {
  template <typename K = TK>
  static ENABLE_IF<std::is_same<std::string, K>::value> Load(rapidjson::Value* source,
                                                             std::map<TK, TV, TC, TA>& destination,
                                                             const std::string& path) {
    if (source && source->IsObject()) {
      destination.clear();
      TK k;
      TV v;
      for (rapidjson::Value::MemberIterator cit = source->MemberBegin(); cit != source->MemberEnd(); ++cit) {
        LoadFromJSONImpl<TK, J>::Load(&cit->name, k, path);
        LoadFromJSONImpl<TV, J>::Load(&cit->value, v, path);
        destination.emplace(k, v);
      }
    } else if (!JSONPatchMode<J>::value || (source && !source->IsObject())) {
      throw JSONSchemaException("map as object", source, path);  // LCOV_EXCL_LINE
    }
  }

  template <typename K = TK>
  static ENABLE_IF<!std::is_same<std::string, K>::value> Load(rapidjson::Value* source,
                                                              std::map<TK, TV, TC, TA>& destination,
                                                              const std::string& path) {
    if (source && source->IsArray()) {
      destination.clear();
      for (rapidjson::Value::ValueIterator cit = source->Begin(); cit != source->End(); ++cit) {
        if (!cit->IsArray()) {
          throw JSONSchemaException("map entry as array", source, path);  // LCOV_EXCL_LINE
        }
        if (cit->Size() != 2u) {
          throw JSONSchemaException("map entry as array of two elements", source, path);  // LCOV_EXCL_LINE
        }
        TK k;
        TV v;
        LoadFromJSONImpl<TK, J>::Load(&(*cit)[static_cast<rapidjson::SizeType>(0)], k, path);
        LoadFromJSONImpl<TV, J>::Load(&(*cit)[static_cast<rapidjson::SizeType>(1)], v, path);
        destination.emplace(k, v);
      }
    } else if (!JSONPatchMode<J>::value || (source && !source->IsArray())) {
      throw JSONSchemaException("map as array", source, path);  // LCOV_EXCL_LINE
    }
  }
};

}  // namespace load
}  // namespace json

namespace binary {
namespace save {

template <typename TK, typename TV, typename TC, typename TA>
struct SaveIntoBinaryImpl<std::map<TK, TV, TC, TA>> {
  static void Save(std::ostream& ostream, const std::map<TK, TV, TC, TA>& value) {
    SaveSizeIntoBinary(ostream, value.size());
    for (const auto& element : value) {
      SaveIntoBinaryImpl<std::pair<TK, TV>>::Save(ostream, element);
    }
  }
};

}  // namespace save

namespace load {

template <typename TK, typename TV, typename TC, typename TA>
struct LoadFromBinaryImpl<std::map<TK, TV, TC, TA>> {
  static void Load(std::istream& istream, std::map<TK, TV, TC, TA>& destination) {
    destination.clear();
    BINARY_FORMAT_SIZE_TYPE size = LoadSizeFromBinary(istream);
    TK k;
    TV v;
    for (size_t i = 0; i < static_cast<size_t>(size); ++i) {
      LoadFromBinaryImpl<TK>::Load(istream, k);
      LoadFromBinaryImpl<TV>::Load(istream, v);
      destination.emplace(k, v);
    }
  }
};

}  // namespace load
}  // namespace binary
}  // namespace serialization
}  // namespace current

#endif  // CURRENT_TYPE_SYSTEM_SERIALIZATION_TYPES_MAP_H
