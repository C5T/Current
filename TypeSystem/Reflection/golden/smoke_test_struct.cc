// g++ -c -std=c++11 current.cc

#include <string>
#include <cstdint>
#include <vector>
#include <map>
#include <utility>

namespace current {
struct Primitives {
  uint8_t a;
  uint16_t b;
  uint32_t c;
  uint64_t d;
  int8_t e;
  int16_t f;
  int32_t g;
  int64_t h;
  char i;
  std::string j;
  float k;
  double l;
  bool m;
};
struct FullTest {
  Primitives primitives;
  std::vector<std::string> v1;
  std::vector<Primitives> v2;
  std::pair<std::string, Primitives> p;
  Optional<Primitives> o;
};
}  // namespace current
