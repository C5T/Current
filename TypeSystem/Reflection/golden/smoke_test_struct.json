{
  "structs": [
    [
      "T9202869030182953577",
      {
        "type_id": "T9202869030182953577",
        "name": "FullTest",
        "super_type_id": "T0",
        "fields": [
          [
            "T9206969065948310524",
            "primitives"
          ],
          [
            "T9319767778871345491",
            "v1"
          ],
          [
            "T9315520306457829347",
            "v2"
          ],
          [
            "T9334519405000386225",
            "p"
          ],
          [
            "T9211849004702662543",
            "o"
          ]
        ]
      }
    ],
    [
      "T9206969065948310524",
      {
        "type_id": "T9206969065948310524",
        "name": "Primitives",
        "super_type_id": "T0",
        "fields": [
          [
            "T9000000000000000021",
            "a"
          ],
          [
            "T9000000000000000022",
            "b"
          ],
          [
            "T9000000000000000023",
            "c"
          ],
          [
            "T9000000000000000024",
            "d"
          ],
          [
            "T9000000000000000031",
            "e"
          ],
          [
            "T9000000000000000032",
            "f"
          ],
          [
            "T9000000000000000033",
            "g"
          ],
          [
            "T9000000000000000034",
            "h"
          ],
          [
            "T9000000000000000041",
            "i"
          ],
          [
            "T9000000000000000042",
            "j"
          ],
          [
            "T9000000000000000051",
            "k"
          ],
          [
            "T9000000000000000052",
            "l"
          ],
          [
            "T9000000000000000011",
            "m"
          ],
          [
            "T9000000000000000061",
            "n"
          ],
          [
            "T9000000000000000062",
            "o"
          ]
        ]
      }
    ]
  ],
  "types": [
    [
      "T9211849004702662543",
      {
        "type_id": "T9211849004702662543",
        "enum_name": "",
        "included_types": [
          "T9206969065948310524"
        ]
      }
    ],
    [
      "T9315520306457829347",
      {
        "type_id": "T9315520306457829347",
        "enum_name": "",
        "included_types": [
          "T9206969065948310524"
        ]
      }
    ],
    [
      "T9319767778871345491",
      {
        "type_id": "T9319767778871345491",
        "enum_name": "",
        "included_types": [
          "T9000000000000000042"
        ]
      }
    ],
    [
      "T9334519405000386225",
      {
        "type_id": "T9334519405000386225",
        "enum_name": "",
        "included_types": [
          "T9000000000000000042",
          "T9206969065948310524"
        ]
      }
    ]
  ],
  "ordered_struct_list": [
    "T9206969065948310524",
    "T9202869030182953577"
  ]
}